File: api/src/test/java/net/blueshell/api/factory/model/BaseFactory.java
package net.blueshell.api.factory.model;

import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.base.BaseModel;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Random;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Base class for model factories.
 */
@RequiredArgsConstructor
public abstract class BaseFactory<T> {

    private static final AtomicLong idCounter = new AtomicLong(1000);
    protected final Faker faker;
    protected final PasswordEncoder passwordEncoder;
    protected final Random random;

    /** Create a basic instance with minimal required fields. */
    public abstract T createBasic();

    /** Create a fully populated instance. */
    public abstract T createFull();

    /** Create an instance with specific customizations. */
    public abstract T createWithCustomizations(java.util.function.Consumer<T> customizer);

    protected Long generateId() {
        return idCounter.incrementAndGet();
    }

    protected Instant futureInstant() {
        return LocalDateTime.now().plusDays(faker.number().numberBetween(1, 365))
                .atZone(ZoneId.systemDefault()).toInstant();
    }

    protected Instant pastInstant() {
        return LocalDateTime.now().minusDays(faker.number().numberBetween(1, 365))
                .atZone(ZoneId.systemDefault()).toInstant();
    }

    protected LocalDate futureDate() {
        return LocalDate.now().plusDays(faker.number().numberBetween(1, 365));
    }

    protected LocalDate pastDate() {
        return LocalDate.now().minusDays(faker.number().numberBetween(1, 365));
    }

    /** Hook to set base entity fields if needed (usually set by JPA). */
    protected <M extends BaseModel> void setBaseFields(M entity) {
        // Intentionally no-op for tests
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/BaseFactory.kt
package net.blueshell.api.factory.model

import com.github.javafaker.Faker
import net.blueshell.api.model.base.BaseModel
import org.springframework.security.crypto.password.PasswordEncoder
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Random
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Base class for model factories.
 */
abstract class BaseFactory<T>(
    protected val faker: Faker,
    protected val passwordEncoder: PasswordEncoder,
    protected val random: Random
) {

    /** Create a basic instance with minimal required fields. */
    abstract fun createBasic(): T

    /** Create a fully populated instance. */
    abstract fun createFull(): T

    /** Create an instance with specific customizations. */
    abstract fun createWithCustomizations(customizer: Consumer<T>): T

    protected fun generateId(): Long = ID_COUNTER.incrementAndGet()

    protected fun futureInstant(): Instant {
        return LocalDateTime.now().plusDays(faker.number().numberBetween(1, 365).toLong())
            .atZone(ZoneId.systemDefault()).toInstant()
    }

    protected fun pastInstant(): Instant {
        return LocalDateTime.now().minusDays(faker.number().numberBetween(1, 365).toLong())
            .atZone(ZoneId.systemDefault()).toInstant()
    }

    protected fun futureDate(): LocalDate = LocalDate.now().plusDays(faker.number().numberBetween(1, 365).toLong())

    protected fun pastDate(): LocalDate = LocalDate.now().minusDays(faker.number().numberBetween(1, 365).toLong())

    /** Hook to set base entity fields if needed (usually set by JPA). */
    protected fun <M : BaseModel> setBaseFields(entity: M) {
        // Intentionally no-op for tests
    }

    private companion object {
        val ID_COUNTER = AtomicLong(1000)
    }
}
